pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go

```diff
@@ -0,0 +1,260 @@
+	lastActivities   *shrinkingmap.ShrinkingMap[account.SeatIndex, time.Time]
+	lastActivityTime time.Time
+	activityMutex    syncutils.RWMutex
+	committeeMutex   syncutils.RWMutex
+
+	optsSeatCount              uint32
+	optsActivityWindow         time.Duration
+	optsOnlineCommitteeStartup []iotago.AccountID
+
+	module.Module
+}
+
+// NewProvider returns a new sybil protection provider that uses the ProofOfStake module.
+func NewProvider(opts ...options.Option[SeatManager]) module.Provider[*engine.Engine, seatmanager.SeatManager] {
+	return module.Provide(func(e *engine.Engine) seatmanager.SeatManager {
+		return options.Apply(
+			&SeatManager{
+				apiProvider:     e,
+				events:          seatmanager.NewEvents(),
+				committeeStore:  e.Storage.Committee(),
+				onlineCommittee: ds.NewSet[account.SeatIndex](),
+				inactivityQueue: timed.NewPriorityQueue[account.SeatIndex](true),
+				lastActivities:  shrinkingmap.New[account.SeatIndex, time.Time](),
+
+				optsActivityWindow: time.Second * 30,
+			}, opts, func(s *SeatManager) {
+				e.Events.SeatManager.LinkTo(s.events)
+
+				e.HookConstructed(func() {
+					s.TriggerConstructed()
+
+					// We need to mark validators as active upon solidity of blocks as otherwise we would not be able to
+					// recover if no node was part of the online committee anymore.
+					e.Events.CommitmentFilter.BlockAllowed.Hook(func(block *blocks.Block) {
+						// Only track identities that are part of the committee.
+						committee, exists := s.CommitteeInSlot(block.ID().Slot())
+						if !exists {
+							panic(ierrors.Errorf("committee not selected for slot %d, but received block in that slot", block.ID().Slot()))
+						}
+
+						seat, exists := committee.GetSeat(block.ProtocolBlock().IssuerID)
+						if exists {
+							s.markSeatActive(seat, block.ProtocolBlock().IssuerID, block.IssuingTime())
+						}
+
+						s.events.BlockProcessed.Trigger(block)
+					})
+				})
+			})
+	})
+}
+
+var _ seatmanager.SeatManager = &SeatManager{}
+
+func (s *SeatManager) RotateCommittee(epoch iotago.EpochIndex, candidates *account.Accounts) (*account.SeatedAccounts, error) {
+	s.committeeMutex.RLock()
+	defer s.committeeMutex.RUnlock()
+
+	type poolWithAccountID struct {
+		accountID iotago.AccountID
+		pool      *account.Pool
+	}
+
+	candidatePools := make([]*poolWithAccountID, 0)
+	candidates.ForEach(func(id iotago.AccountID, pool *account.Pool) bool {
+		candidatePools = append(candidatePools, &poolWithAccountID{pool: pool, accountID: id})
+
+		return true
+	})
+
+	sort.Slice(candidatePools, func(i, j int) bool {
+		if candidatePools[i].pool.PoolStake != candidatePools[j].pool.PoolStake {
+			return candidatePools[i].pool.PoolStake > candidatePools[j].pool.PoolStake
+		}
+
+		if candidatePools[i].pool.ValidatorStake != candidatePools[j].pool.ValidatorStake {
+			return candidatePools[i].pool.ValidatorStake > candidatePools[j].pool.ValidatorStake
+		}
+
+		// TODO: add more tie-breaking
```

Comment:
I added some more things there. It wasn't there initially because the data wasn't available. Now I changed what RotateCommittee receives which imo makes it clearer.